bqskit/passes/processing/treescan.py

```diff
@@ -0,0 +1,226 @@
+"""This module implements the ScanningGateRemovalPass."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+from typing import Callable
+
+from bqskit.compiler.basepass import BasePass
+from bqskit.compiler.passdata import PassData
+from bqskit.ir.circuit import Circuit
+from bqskit.ir.operation import Operation
+from bqskit.ir.opt.cost.functions import HilbertSchmidtResidualsGenerator
+from bqskit.ir.opt.cost.generator import CostFunctionGenerator
+from bqskit.runtime import get_runtime
+from bqskit.utils.typing import is_integer
+from bqskit.utils.typing import is_real_number
+
+_logger = logging.getLogger(__name__)
+
+
+class TreeScanningGateRemovalPass(BasePass):
+    """
+    The ScanningGateRemovalPass class.
+
+    Starting from one side of the circuit, run the following:
+
+    Split the circuit operations into chunks of size tree_depth
+    At every iteration:
+    a. Look at the next chunk of operations
+    b. Generate 2 ^ tree_depth circuits. Each circuit corresponds to every
+    combination of whether or not to include one of the operations in the chunk.
+    c. Instantiate in parallel all 2^tree_depth circuits
+    d. Choose the circuit that has the least number of operations and move
+    on to the next chunk of operations.
+
+    This optimization is less greedy than the current ScanningGate removal,
```

Comment:
For better documentation, you may want to link ScanningGateRemoval using the proper syntax:

```raw
:class:`~bqskit.passes.processing.ScanningGateRemovalPass`
```
We can also double check the documentation before merging by clicking on the PR readthedocs test.